### PR TITLE
GP2HUB-121 Related Ouputs in Output Detail Page

### PR DESCRIPTION
**Pull request by gabiayako**
Jira ticket: https://asaphub.atlassian.net/browse/GP2HUB-121

---

This PR adds the related outputs card to the output detail page. In order to do that, it was necessary to change the output query so it also brings the name of the source (projects/working groups) and its id.

---

Mobile | Desktop   
:-------------------------:|:-------------------------:
  |

### CODE REVIEW

File: packages/gp2-components/src/utils/icon.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      return outputMaterial;
+    default:
+      return outputReport;
+  }
+};
+
+export const getSourceIcon = (source: 'Project' | 'Working Group') => {
+  if (source === 'Project') return projectIcon;
+  return workingGroupIcon;
+};
```
Comment (markcsims) [inline packages/gp2-components/src/utils/icon.ts:37]:
```suggestion
export const getSourceIcon = (source: 'Project' | 'Working Group') => 
  source === 'Project' ? projectIcon : workingGroupIcon;
```

File: packages/react-components/src/organisms/RelatedResearchCard.tsx [hunk 2]
```diff
@@ -99,14 +84,35 @@ const showMoreStyles = css({
...
 const RelatedResearchCard: React.FC<RelatedResearchCardProp> = ({
   relatedResearch,
   description,
+  getSourceIcon,
+  title = 'Related Research',
+  getIconForDocumentType = getIconForDocumentTypeCRN,
+  tableTitles = [
+    'Document Type',
+    'Shared Output Name',
+    'Team or Working Group',
```
Comment (markcsims) [inline packages/react-components/src/organisms/RelatedResearchCard.tsx:114]: I'm a bit confused by this > 

The title is "Team or Working Group" - but on line 101 we can have Project or Working Group

Comment (markcsims) [reply packages/react-components/src/organisms/RelatedResearchCard.tsx:114]: I couldn't see it anywhere, but are we mixing things. i.e. GP2 has Project and Working Group but CRN has Team and Working Group?

File: packages/react-components/src/organisms/RelatedResearchCard.tsx [hunk 3]
```diff
@@ -120,73 +126,107 @@ const RelatedResearchCard: React.FC<RelatedResearchCardProp> = ({
...
+              ) : (
+                output.entity &&
+                getSourceIcon && (
+                  <div css={paragraphStyle}>
+                    <div css={iconStyles}>
+                      {getSourceIcon(
+                        output.entity?.type === 'workingGroups'
+                          ? 'Working Group'
+                          : 'Project',
+                      )}
```
Comment (markcsims) [inline packages/react-components/src/organisms/RelatedResearchCard.tsx:211]: We could change getSourceIcon to accept the entity.type as the property?

File: packages/react-components/src/organisms/__tests__/RelatedResearchCard.test.tsx [hunk 4]
```diff
@@ -32,6 +32,26 @@ it('displays the related research card header and description', () => {
...
+      ]}
+      title="Related Outputs"
+      description="Find out all outputs that contributed to this one."
+    />,
+  );
+  expect(getByRole('heading', { name: 'Related Outputs' })).toBeVisible();
+  expect(
+    getByText('Find out all outputs that contributed to this one.'),
+  ).toBeVisible();
+  expect(queryByText('View More Outputs')).toBeNull();
```
Comment (markcsims) [inline packages/react-components/src/organisms/__tests__/RelatedResearchCard.test.tsx:52]: Shouldn't it be
```
expect(screen.queryByText('View More Outputs')).not.toBeInTheDocument()
```

File: packages/gp2-components/src/utils/icon.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  outputCode,
+  outputDataset,
+  outputForm,
+  outputMaterial,
+  outputReport,
+  projectIcon,
+  workingGroupIcon,
+} from '../icons';
+
+export const getIconForDocumentType = (
```
Comment (markcsims) [inline packages/gp2-components/src/utils/icon.ts:13]: This function also exists here, maybe we should refactor to have only one:

https://github.com/yldio/asap-hub/blob/813c79f09256a152ee3d37f46bbff0094c8be387/packages/gp2-components/src/organisms/OutputRelatedResearchCard.tsx#L17-L36

File: packages/react-components/src/utils/icon.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    case 'Bioinformatics':
+      return bioinformatics;
+    case 'Lab Resource':
+      return labResource;
+    case 'Grant Document':
+      return grantDocument;
+    default:
+      return protocol;
+  }
+};
```
Comment (markcsims) [inline packages/react-components/src/utils/icon.ts:31]:
```
const icons: Record<ResearchOutputDocumentType, EmotionJSX.Element> = {
  Protocol: protocol,
  Article: article,
  Dataset: dataset,
  Bioinformatics: bioinformatics,
  'Lab Resource': labResource,
  'Grant Document': grantDocument,
  Presentation: protocol,
  Report: protocol,
};

export const getIconForDocumentType = (
  documentType: ResearchOutputDocumentType,
): EmotionJSX.Element => icons[documentType];
```

Comment (markcsims) [reply packages/react-components/src/utils/icon.ts:31]: This we refactors from this lookup, (which I prefer to a switch statement).
But also has 2 extra properties: Presentation and Report.

**Review by markcsims — CHANGES_REQUESTED** [review]

File: packages/gp2-components/src/utils/icon.ts (hunk 5 above)
Comment (gabiayako) [reply packages/gp2-components/src/utils/icon.ts:13]: I've done that here https://github.com/yldio/asap-hub/pull/3779/files#diff-404baf71a79662d27a9e9b4f2ef852c3bd91aeb57a97938be56ad04c6eee2408 but this is still in PO so it hasn't been merged yet

File: packages/react-components/src/organisms/RelatedResearchCard.tsx (hunk 2 above)
Comment (gabiayako) [reply packages/react-components/src/organisms/RelatedResearchCard.tsx:119]: This is
```
[
    'Document Type',
    'Shared Output Name',
    'Team or Working Group',
]
```

the default when I'm not passing any `tableTitles` to the component :) I'm passing the appropriate titles for GP2 here https://github.com/yldio/asap-hub/blob/GP2HUB-121-create-related-outputs-card/packages/gp2-components/src/templates/OutputDetailPage.tsx#L113

so they look like this for GP2

Mobile | Desktop   
:-------------------------:|:-------------------------:
  |

File: packages/react-components/src/organisms/RelatedResearchCard.tsx (hunk 3 above)
Comment (gabiayako) [reply packages/react-components/src/organisms/RelatedResearchCard.tsx]: committed here [review: getSourceIcon suggestion](https://github.com/yldio/asap-hub/pull/3816/commits/a7369f86600cd3b213734eb29c8e69f8bc275eb5)

File: packages/react-components/src/organisms/__tests__/RelatedResearchCard.test.tsx (hunk 4 above)
Comment (gabiayako) [reply packages/react-components/src/organisms/__tests__/RelatedResearchCard.test.tsx:53]: committed here [review: change to be null to not to be in the document](https://github.com/yldio/asap-hub/pull/3816/commits/ae9cdd871751f2c054feac65a0ca7256e9787887)

File: packages/react-components/src/utils/icon.ts (hunk 6 above)
Comment (gabiayako) [reply packages/react-components/src/utils/icon.ts:22]: committed here [review: refactor getIconForDocumentType](https://github.com/yldio/asap-hub/pull/3816/commits/7a42a73e73570b5d4122a09a1c6649f9c57d30d1)

I had some Typescript issues though. When passing a default value for `getIconForDocumentType` like 

```
const RelatedResearchCard: React.FC<RelatedResearchCardProp> = ({
  relatedResearch,
  description,
  getSourceIcon,
  getIconForDocumentType = getIconForDocumentTypeCRN,
  title = 'Related Research',
  tableTitles = [
    'Document Type',
    'Shared Output Name',
    'Team or Working Group',
  ],
}) => {
```

Typescript somehow couldn't accept that `getIconForDocumentType` could receive a gp2.OutputDocumentType documentType so I kept getting this error

```
Type '"Code/Software"' is not assignable to type '"Grant Document" | "Presentation" | "Dataset" | "Bioinformatics" | "Protocol" | "Lab Resource" | "Article" | "Report"'
```

Then I removed the default value and typed `getIconForDocumentType` as 

```
getIconForDocumentType:
  | ((documentType: ResearchOutputDocumentType) => EmotionJSX.Element)
  | ((documentType: gp2.OutputDocumentType) => EmotionJSX.Element);
```

but then Typescript gives this error

```
 Argument of type '"Dataset" | "Article" | "Code/Software" | "GP2 Reports" | "Procedural Form" | "Training Materials"' is not assignable to parameter of type '"Dataset" | "Article"'.
```

when I try to call `getIconForDocumentType` as if it only accepts the intersection of the union types. So the only way I've found to not give an error was to call getIconForDocumentType casting documentType as any. This is a shame, but I haven't found another way to do it. I'm very pleased to accept suggestions though.

**Review by markcsims — APPROVED** [review]

**Comment by gabiayako** [comment]
rebased to master and solved conflicts due to https://github.com/yldio/asap-hub/pull/3779

**Pull request merged**